mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowHttpCaller.java

```diff
@@ -30,20 +30,65 @@
 class MlflowHttpCaller {
   private static final Logger logger = LoggerFactory.getLogger(MlflowHttpCaller.class);
   private static final String BASE_API_PATH = "api/2.0/preview/mlflow";
-  private HttpClient httpClient;
+  protected HttpClient httpClient;
   private final MlflowHostCredsProvider hostCredsProvider;
+  private final int maxRateLimitIntervalMillis;
+  private final int rateLimitRetrySleepInitMillis;
 
   MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider) {
+    this(hostCredsProvider, 60000);
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider, int maxRateLimitIntervalSeconds) {
+    this(hostCredsProvider, maxRateLimitIntervalSeconds, 1000);
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider,
+                   int maxRateLimitIntervalSeconds,
+                   int rateLimitRetrySleepInitMs) {
     this.hostCredsProvider = hostCredsProvider;
+    this.maxRateLimitIntervalMillis = maxRateLimitIntervalSeconds;
+    this.rateLimitRetrySleepInitMillis = rateLimitRetrySleepInitMs;
+  }
+
+  MlflowHttpCaller(MlflowHostCredsProvider hostCredsProvider,
+                   int maxRateLimitIntervalSeconds,
+                   int rateLimitRetrySleepInitMs,
+                   HttpClient client) {
+    this(hostCredsProvider, maxRateLimitIntervalSeconds, rateLimitRetrySleepInitMs);
+    this.httpClient = client;
+  }
+
+
+  HttpResponse executeRequest(HttpRequestBase request) throws IOException {
+    HttpResponse response = httpClient.execute(request);
+    int timeLeft = maxRateLimitIntervalMillis;
+    int sleepFor = rateLimitRetrySleepInitMillis;
+    while (response.getStatusLine().getStatusCode() == 429 && timeLeft > 0) {
```

Comment:
IMHO `do { ... } while()` might work better here, esp if we want to retries for other error conditions like `404` or `503` (perhaps)

```
   HttpResponse response;
   int timeLeft = maxRateLimitIntervalMillis;
   int sleepFor = rateLimitRetrySleepInitMillis;
   do {
      response = httpClient.execute(request);
      if (response.getStatusLine().getStatusCode() == 429) {
        logger.warn("Request returned with status code 429 (Rate limit exceeded). Retrying after "
                    + sleepFor
                    + " milliseconds. Will continue to retry 429s for up to "
                    + timeLeft
                    + " milliseconds.");
        try {
          Thread.sleep(sleepFor);
        } catch (InterruptedException e) {
          logger.warn("interrupted while waiting for retry.");
        }
        sleepFor = Math.min(timeLeft, 2 * sleepFor);
        timeLeft -= sleepFor;
      } else {
        break;
      }
    } while (timeLeft > 0);
```

Reply:
Discussed offline, decided to keep this as is.